### PR TITLE
Compare category bitmaps

### PR DESCRIPTION
**Pull request by BoltzmannBrain**

### CODE REVIEW

**Comment by BoltzmannBrain** [comment]
@scottpurdy please review.

File: fluent/models/classification_model.py [hunk 1]
```diff
@@ -79,6 +79,34 @@ def logEncodings(self, patterns, path):
       f.write(json.dumps(jsonPatterns, indent=1))
 
 
+  def logCategories(self, path, comparisons=None, labelRefs=None):
```
Comment (scottpurdy) [inline fluent/models/classification_model.py]: Some terminology is confusing here. Logging has a very specific meaning so I might call this "writeCategoriesToFile" or something. And path is usually referring to a file where a directory would be called "dir" or "dirName".

File: fluent/models/classification_model.py [hunk 2]
```diff
@@ -79,6 +79,34 @@ def logEncodings(self, patterns, path):
...
+    """
+    For models which define categories with bitmaps, log the categories (and
+    their relative distances) to a JSON specified by path (must be absolute).
```
Comment (scottpurdy) [inline fluent/models/classification_model.py]: Can you specify the format of the files? And does it really have to be absolute or can the caller decide whether they want an absolute or relative path?

File: fluent/models/classification_model.py [hunk 3]
```diff
@@ -79,6 +79,34 @@ def logEncodings(self, patterns, path):
...
+
+    if not os.path.isdir(path):
+      raise ValueError("Invalid path to write file.")
+
+    with open(os.path.join(path, "category_distances.json"), "w") as f:
+      if labelRefs:
+        f.write(json.dumps({i: label for i, label in enumerate(labelRefs)},
+                          sort_keys=False,
+                          indent=4,
+                          separators=(',', ': ')))
```
Comment (scottpurdy) [inline fluent/models/classification_model.py]: double strings

File: fluent/models/classification_model.py [hunk 4]
```diff
@@ -79,6 +79,34 @@ def logEncodings(self, patterns, path):
...
+    if not hasattr(self, "categoryBitmaps"):
+      raise TypeError("This model does not have category encodings compatible "
+                      "for logging.")
+
+    if not os.path.isdir(path):
+      raise ValueError("Invalid path to write file.")
+
+    with open(os.path.join(path, "category_distances.json"), "w") as f:
+      if labelRefs:
+        f.write(json.dumps({i: label for i, label in enumerate(labelRefs)},
```
Comment (scottpurdy) [inline fluent/models/classification_model.py]: `json.dump` (without the trailing "s") takes a file and writes to it so you don't need the `f.write` call

File: fluent/models/classification_model.py [hunk 5]
```diff
@@ -79,6 +79,34 @@ def logEncodings(self, patterns, path):
...
+                          sort_keys=False,
+                          indent=4,
+                          separators=(',', ': ')))
+      if comparisons:
+        f.write(json.dumps(comparisons,
+                          sort_keys=False,
```
Comment (scottpurdy) [inline fluent/models/classification_model.py]: indentation

File: fluent/models/classification_model.py [hunk 6]
```diff
@@ -79,6 +79,34 @@ def logEncodings(self, patterns, path):
...
+                      "for logging.")
+
+    if not os.path.isdir(path):
+      raise ValueError("Invalid path to write file.")
+
+    with open(os.path.join(path, "category_distances.json"), "w") as f:
+      if labelRefs:
+        f.write(json.dumps({i: label for i, label in enumerate(labelRefs)},
+                          sort_keys=False,
+                          indent=4,
```
Comment (scottpurdy) [inline fluent/models/classification_model.py]: 2 spaces would be a little more concise (I think that's what we normally use)

File: fluent/models/classification_model.py (hunk 4 above)
Comment (scottpurdy) [inline fluent/models/classification_model.py]: This could be:

`json.dump(dict(enumerate(labelRefs)))`

File: fluent/models/classify_endpoint.py [hunk 7]
```diff
@@ -149,6 +149,70 @@ def testModel(self, sample, numLabels=3, metric="overlappingAll"):
...
 
   @staticmethod
+  def compareCategories(catDistances, metric="overlappingAll"):
+    """
+    Calculate category distances. Returns a defaultdict of category keys, where
+    values are OrderedDicts sorted such that the most similar categories
+    (according to the input metric) are listed first.
+    """
+    descendingOrder = ("overlappingAll", "overlappingLeftRight",
+        "overlappingRightLeft", "cosineSimilarity", "weightedScoring")
```
Comment (scottpurdy) [inline fluent/models/classify_endpoint.py]: please use one of the indentation options listed here: https://www.python.org/dev/peps/pep-0008/#indentation

File: fluent/models/classify_endpoint.py [hunk 8]
```diff
@@ -149,6 +149,70 @@ def testModel(self, sample, numLabels=3, metric="overlappingAll"):
...
+
+    categoryComparisons = defaultdict(list)
+    for k, v in catDistances.iteritems():
+      # Create a dict for this category
+      metricDict = {compareCat: distances[metric]
+                    for compareCat, distances in v.iteritems()}
+      # Sort the dict by the metric
+      if metric in descendingOrder:
+        categoryComparisons[k] = OrderedDict(
+            sorted(metricDict.items(), key=lambda k: k[1])[::-1])
```
Comment (scottpurdy) [inline fluent/models/classify_endpoint.py]: `sorted` has a reverse option so you don't have to reverse it after the sort

File: tests/unit/classification_model_test.py [hunk 9]
```diff
@@ -137,6 +137,50 @@ def testClassifyRandomSDRSingleAndMultiClass(self):
...
+
+    # Assert correct sorting of categories for different metrics.
+    expected = {
+        0: OrderedDict([(0, 3), (2, 1), (1, 0)]),
+        1: OrderedDict([(1, 3), (2, 2), (0, 0)]),
+        2: OrderedDict([(2, 3), (1, 2), (0, 1)]),
+        }
+    catComparisons = model.compareCategories(
+        catDistances, metric="overlappingAll")
+    self.assertEqual(expected, catComparisons,
```
Comment (scottpurdy) [inline tests/unit/classification_model_test.py]: `assertDictEquals` would give a little extra specificity

**Comment by scottpurdy** [comment]
done reviewing

**Comment by BoltzmannBrain** [comment]
Thank you @scottpurdy, I made changes in response to your comments.

File: fluent/models/classification_model.py [hunk 10]
```diff
@@ -75,10 +75,47 @@ def logEncodings(self, patterns, path):
...
+      raise ValueError("Invalid path to write file.")
+
+    with open(os.path.join(dirName, "category_distances.json"), "w") as f:
+      if labelRefs:
+        json.dump(
+            dict(enumerate(labelRefs)),
+            f,
+            sort_keys=False,
+            indent=2,
+            separators=(',', ': '))
```
Comment (scottpurdy) [inline fluent/models/classification_model.py]: double quotes

File: tests/unit/classification_model_test.py [hunk 11]
```diff
@@ -137,6 +137,50 @@ def testClassifyRandomSDRSingleAndMultiClass(self):
                     "Outputs for samples 2 and 4 should be identical.")
 
 
+  def testCompareCategories(self):
+    model =ClassificationModelEndpoint()
```
Comment (scottpurdy) [inline tests/unit/classification_model_test.py]: space after `=`

**Comment by scottpurdy** [comment]
a couple nit picky comments but otherwise :+1:

**Pull request merged**